File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int num) {
        if (head == null) return null;
        int counter = 0;
        ListNode currNode = head;
        while(currNode != null) {
            currNode = currNode.next;
            counter++;
        }
        if (num == counter) {
            return head.next;
        }
        currNode = head;
        for (int i = 1; i < counter-num; i++) {
            currNode = currNode.next;
        }
        if (currNode.next != null) {
            currNode.next = currNode.next.next;
        }
        return head;
    }}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prev_group = dummy

        while True:
            kth = prev_group
            for _ in range(k):
                kth = kth.next

                if not kth:
                    return dummy.next
                
            next_group = kth.next
            prev = next_group
            curr = prev_group.next

            while(curr != next_group):
                temp = curr.next
                curr.next = prev
                prev = curr
                curr = temp
            
            group_start = prev_group.next
            prev_group.next = kth
            prev_group = group_start
        